core/src/main/scala/com/salesforce/op/utils/stages/NameDetectUtils.scala

```diff
@@ -0,0 +1,337 @@
+  }
+
+  private[op] def dictCheck(tokens: Seq[String], dict: NameDictionary): Double = {
+    if (tokens.isEmpty) 0.0 else {
+      tokens.map({ token: String => if (dict contains token) 1 else 0}).sum.toDouble / tokens.length
+    }
+  }
+
+  private[op] def genderDictCheck(nameToCheckGenderOf: String, genderDict: GenderDictionary): String = {
+    genderDict.get(nameToCheckGenderOf).map(
+      probMale => if (probMale >= 0.5) Male else Female
+    ).getOrElse(GenderNA)
+  }
+
+  private[op] def identifyGender(
+    text: String,
+    tokens: Seq[String],
+    strategy: GenderDetectStrategy,
+    genderDict: GenderDictionary
+  ): String = {
+    strategy match {
+      case FindHonorific() =>
+        val matched = tokens filter { NameDetectUtils.AllHonorifics contains }
+        if (matched.length == 1) {
+          if (MaleHonorifics contains matched.head) Male else Female
+        }
+        else GenderNA
+      case ByIndex(index) =>
+        val nameToCheckGenderOf = tokens.lift(index).getOrElse("")
+        genderDictCheck(nameToCheckGenderOf, genderDict)
+      case ByLast() =>
+        val nameToCheckGenderOf = tokens.lastOption.getOrElse("")
+        genderDictCheck(nameToCheckGenderOf, genderDict)
+      case ByRegex(pattern) =>
+        text match {
+          case pattern(matchedGroup) =>
+            val nameToCheckGenderOf = preProcess(Some(matchedGroup)).headOption.getOrElse("")
+            genderDictCheck(nameToCheckGenderOf, genderDict)
+          case _ => GenderNA
+        }
+      case _ =>
+        logError("Unimplemented gender detection strategy found")
+        GenderNA
+    }
+  }
+
+  private[op] def computeGenderResultsByStrategy(
+    text: String,
+    tokens: Seq[String],
+    genderDict: GenderDictionary
+  ): Map[String, GenderStats] = {
+    GenderDetectStrategies map { strategy: GenderDetectStrategy =>
+      val genderResult: String = identifyGender(text, tokens, strategy, genderDict)
+      strategy.toString -> GenderStats(
+        if (genderResult == Male) 1 else 0,
+        if (genderResult == Female) 1 else 0,
+        if (genderResult == GenderNA) 1 else 0
+      )
+    } toMap
+  }
+
+  private[op] def computeResults(
+    input: String,
+    nameDict: Broadcast[NameDictionary],
+    genderDict: Broadcast[GenderDictionary],
+    hll: HyperLogLogMonoid
+  ): NameDetectStats = input match {
+    case "" if $(ignoreNulls) => NameDetectStats.empty
+    case _ =>
+      val tokens = preProcess(input)
+      NameDetectStats(
+        computeGuardCheckQuantities(input, tokens, hll),
+        AveragedValue(1L, dictCheck(tokens, nameDict.value)),
+        computeGenderResultsByStrategy(input, tokens, genderDict.value)
+      )
+  }
+
+  type NameDetectMapFun[T <: Text] = T#Value => NameDetectStats
+  private[op] def makeMapFunction[T <: Text](spark: SparkSession): NameDetectMapFun[T] = {
+    val broadcastNameDict: Broadcast[NameDictionary] = spark.sparkContext.broadcast(DefaultNameDictionary)
```

Comment:
what are the performance implications of the broadcast variable here? why not use `object` instead?

Reply:
This seems like the clearest use case for Spark's broadcast variables (a large read-only Map to use in RDD or dataset operations). Without it, Spark would serialize/deserialize the name dictionaries for each worker node. Locally, it does make tests slower, and probably for our many small-to-medium size orgs we wouldn't see that much of a difference. But for those XL orgs it probably makes a difference. Open to other thoughts though; it doesn't look like TMOG uses broadcast variables at any other place.

Reply:
It was definitely a headache to figure out how to write the name detection code in a shared utils file while also using broadcast variables so there are some code maintainability gains to be had. :)